vernac/mltop.ml

```diff
@@ -397,7 +427,8 @@ let declare_ml_modules local mnames =
   let mnames = List.map (PluginSpec.of_package ~usercode:true) mnames in
   if Lib.sections_are_opened()
   then CErrors.user_err Pp.(str "Cannot Declare ML Module while sections are opened.");
-  let mdigests = CList.concat_map PluginSpec.digest mnames in
+  let mnames = PluginSpec.add_deps mnames in
+  let mdigests = CList.concat_map (fun (_,plugin) -> PluginSpec.digest plugin) mnames in
```

Comment:
Also since we're not doing anything cryptographic and we only care about the set of plugins used, we could simply concatenate the digests and hash the result so that we get a representation that is constant length instead of linear in the number of plugins. This would be just as safe.

Reply:
(The XOR of all digests is both constant-length and order invariant, btw.)

Reply:
That seems over-optimization to me, as the space overhead is O(1) and pretty small in practice.

Tho, the only thing we'd lose by doing this would be the ability to report the concrete plugin the has mismatch happened. This seems like a useful feature.

Reply:
> the space overhead is O(1)

No, it's linear in the number of OCaml dependencies. I agree it probably doesn't matter in practice but in theory with a linear chain of plugins we could get a quadratic overhead.

> report the concrete plugin the has mismatch happened

It's not clear to me from the code which part is responsible for signalling that there is a plugin mismatch. The current PR is adding this data for the build system to realize that something changed plugin-wise at the file level, but AFAIU it doesn't use it to error out. Who is doing that, is it some Coq-facing code or is it a Findlib primitive?

Reply:
Nothing checks for mismatch, it's a hypothetical feature

Reply:
>No, it's linear in the number of OCaml dependencies. I agree it probably doesn't matter in practice but in theory with a linear chain of plugins we could get a quadratic overhead.

No, we filter out already-linked plugins. So eg
~~~
Declare ML Module "coq-core.plugins.cc_core".
Declare ML Module "coq-core.plugins.cc".
~~~
the first one gets the digest for cc_core and any not yet linked dependencies
the second one gets the digest for cc but not cc_core

Reply:
Oh right, we don't do the checking for plugins... Indeed the space issue is more O(k*n) where k is small and usually fixed, so IMHO space is not a worry. For the case of non-depending plugins, overhead is 0 IIUC.

So I guess whatever we do is OK then, but keeping a list of hashes seems simpler to me (premature optimization otherwise?)

Reply:
I'm fine with the current situation, the performance issue was more some theoretical nitpicking. I'm vaguely more concerned about the potential fact that findlib returns different topological orderings depending on the phase of the moon, but it shouldn't matter much in practice either.